File: packages/shared/src/stages.ts
export const ALL_STAGES = [
  'briefings',
  'ready',
  'testing',
  'implementing',
  'review',
  'probing',
  'staged',
  'done',
  'blocked',
] as const;

export type StageId = (typeof ALL_STAGES)[number];

export const TRANSITION_MATRIX: Record<StageId, readonly StageId[]> = {
  briefings: ['ready', 'blocked'],
  ready: ['testing', 'implementing', 'probing', 'blocked', 'briefings'],
  testing: ['implementing', 'blocked'],
  implementing: ['review', 'blocked'],
  probing: ['ready', 'staged', 'blocked'],
  review: ['testing', 'implementing', 'probing', 'blocked'],
  staged: ['done', 'testing', 'implementing', 'ready', 'blocked'],
  done: [],
  blocked: ['ready'],
};

/**
 * Answers "may an item move from `from` to `to`?".
 *
 * `from` is typed as StageId but in practice arrives as unvalidated DB data
 * (Item.stageId is a plain String column, not an enum), so an unknown or
 * legacy stage id would index the matrix to `undefined` and throw a
 * TypeError — surfacing as a 500 from POST /api/board/move instead of a
 * clean "invalid transition" 400. An unrecognized origin stage has no legal
 * transitions, so it answers false.
 */
export function isValidTransition(from: StageId, to: StageId): boolean {
  return TRANSITION_MATRIX[from]?.includes(to) ?? false;
}

export function getValidNextStages(from: StageId): readonly StageId[] {
  return TRANSITION_MATRIX[from];
}

/**
 * Terminal stages that satisfy a dependency: a dependent item is unblocked
 * once every dependency reaches one of these stages.
 *
 * 'staged' counts alongside 'done' because the per-item pipeline now ends
 * at 'staged' — nothing reaches 'done' until Frankie's mission-tail walk
 * promotes it, so gating dependency waves on 'done' alone would deadlock
 * every multi-wave mission. See adr/0005-done-is-terminal-no-in-mission-rework.md.
 */
const DEPENDENCY_SATISFYING_STAGES: readonly StageId[] = ['staged', 'done'];

/**
 * Answers "is a dependency in this stage satisfied?" — true for 'staged'
 * and 'done', false for every other stage. Use this everywhere dependency
 * completion is checked instead of comparing against 'done' directly, so
 * every call site stays in lockstep (see WI-788).
 */
export function isDependencySatisfied(stageId: StageId): boolean {
  return DEPENDENCY_SATISFYING_STAGES.includes(stageId);
}

/**
 * Maps each pipeline stage to the agent responsible for it and the
 * expected next stage in the happy-path pipeline flow.
 *
 * Used by the MCP server to build actionable error messages that tell
 * the orchestrator exactly which agent to dispatch when a transition
 * is rejected.
 */
export interface PipelineStageInfo {
  /** Agent responsible for work in this stage */
  readonly agent: string;
  /** Display name shown in error messages */
  readonly agentDisplay: string;
  /** The expected next stage in the happy-path pipeline */
  readonly nextStage: StageId | null;
  /** Human-readable description of what happens in this stage */
  readonly description: string;
}

export const PIPELINE_STAGES: Partial<Record<StageId, PipelineStageInfo>> = {
  testing: {
    agent: 'murdock',
    agentDisplay: 'Murdock',
    nextStage: 'implementing',
    description: 'writes tests defining acceptance criteria',
  },
  implementing: {
    agent: 'ba',
    agentDisplay: 'B.A.',
    nextStage: 'review',
    description: 'implements code to pass tests',
  },
  review: {
    agent: 'lynch',
    agentDisplay: 'Lynch',
    nextStage: 'probing',
    description: 'reviews tests and implementation together',
  },
  probing: {
    agent: 'amy',
    agentDisplay: 'Amy',
    nextStage: 'staged',
    description: 'investigates for bugs beyond test coverage',
  },
};

File: packages/ateam-cli/internal/stages/stages.go
// Package stages mirrors the stage transition rules and pipeline agent mapping
// defined in packages/shared/src/stages.ts.
//
// The contract test in stages_test.go loads packages/shared/fixtures/stage-matrix.json
// (exported by the shared-package build script) and verifies that ValidTransitions
// and PipelineStages match the TypeScript source exactly. Any drift between the
// two runtimes fails CI here, not at runtime.
package stages

// PipelineStageInfo describes a single pipeline stage: which agent is
// responsible, the display name used in messages, the expected next stage on
// the happy path, and a human-readable description.
//
// The field names and types are pinned by a compile-time contract in
// stages_test.go — rename or retype any field and the package will fail to
// build.
type PipelineStageInfo struct {
	Agent        string
	AgentDisplay string
	NextStage    string
	Description  string
}

// ValidTransitions is the Go mirror of TRANSITION_MATRIX from
// packages/shared/src/stages.ts. Each key is a stage ID; the slice contains
// every legal destination stage for that source stage.
//
// Target slices are treated as sets by the contract test — order does not
// matter, but membership must match exactly.
var ValidTransitions = map[string][]string{
	"briefings":    {"ready", "blocked"},
	"ready":        {"testing", "implementing", "probing", "blocked", "briefings"},
	"testing":      {"implementing", "blocked"},
	"implementing": {"review", "blocked"},
	"probing":      {"ready", "done", "blocked"},
	"review":       {"testing", "implementing", "probing", "blocked"},
	"done":         {},
	"blocked":      {"ready"},
}

// PipelineStages is the Go mirror of PIPELINE_STAGES from
// packages/shared/src/stages.ts. Only stages that have an assigned pipeline
// agent are included (briefings, ready, done, and blocked have no agent and
// are intentionally absent).
//
// All four fields (Agent, AgentDisplay, NextStage, Description) must match
// the TypeScript source character-for-character; the contract test checks each
// one independently and reports which field drifted.
var PipelineStages = map[string]PipelineStageInfo{
	"testing": {
		Agent:        "murdock",
		AgentDisplay: "Murdock",
		NextStage:    "implementing",
		Description:  "writes tests defining acceptance criteria",
	},
	"implementing": {
		Agent:        "ba",
		AgentDisplay: "B.A.",
		NextStage:    "review",
		Description:  "implements code to pass tests",
	},
	"review": {
		Agent:        "lynch",
		AgentDisplay: "Lynch",
		NextStage:    "probing",
		Description:  "reviews tests and implementation together",
	},
	"probing": {
		Agent:        "amy",
		AgentDisplay: "Amy",
		NextStage:    "done",
		Description:  "investigates for bugs beyond test coverage",
	},
}
